### PR TITLE
Compute shares

### PR DESCRIPTION
**Pull request by liamsi**
## Description

- [x] add a way to compute shares (done here but could be simpler though)
- [ ] feed these shares into rsmt2d lib (wip in #63)
- [ ] use resulting extended square to feed this into the NMT and to compute row & col roots (wip in #63)

Closes: #XXX

### CODE REVIEW

File: types/shares.go [hunk 1]
```diff
@@ -0,0 +1,148 @@
+package types
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+
+	"github.com/gogo/protobuf/proto"
```
Comment (marbar3778) [inline types/shares.go:8]: just a minor note: I would be wary of adding gogoproto further into the repo. The upstream repo will work on removing this dependency in the future.

File: types/shares.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+			rawShare := rawData
+			paddedShare := zeroPadIfNecessary(rawShare, shareSize)
+			share := NamespacedShare{paddedShare, nid}
+			shares = append(shares, share)
+		} else { // len(rawData) >= shareWithoutNidSize:
+			shares = append(shares, split(rawData, shareSize, nid)...)
+		}
+	}
+	return shares
+}
```
Comment (liamsi) [inline types/shares.go:116]: @adlerjohn I'll leave away prefixing with the reserved share bytes and initially treat Tx, Evidence etc no different than Messages in the sense that one share will contain one Tx.

File: types/shares.go (hunk 1 above)
Comment (liamsi) [reply types/shares.go:8]: Ah OK, thanks! As soon as upstream will remove this, we'll upgrade too. Here, we only use it to "decorate" (gogo) proto.Messages with a `MarshalDelimited` method. For Tx, and (LL) Message, I have to length prefix manually anyways though.

File: types/shares.go (hunk 2 above)
Comment (liamsi) [reply types/shares.go:116]: Meaning, I'm intentionally skipping the `*` in https://github.com/lazyledger/lazyledger-specs/blob/master/specs/figures/share.svg as there is always only one Tx per share in the first version.

**Pull request closed without merging**